django/contrib/admindocs/utils.py

```diff
@@ -142,3 +142,96 @@ def default_reference_role(name, rawtext, text, lineno, inliner, options=None, c
 
     for name, urlbase in ROLES.items():
         create_reference_role(name, urlbase)
+
+named_group_beg_matcher = re.compile(r'\(\?P(<\w+>)')  # Matches the beginning of a named group
+non_named_group_beg_matcher = re.compile(r'(?<!\\\)\((?!\?P<\w+>))')  # Matches the beginning of an unnamed group
+
+
+def replace_named_groups(pattern):
+    """
+    Find named groups in the pattern supplied, and replace it with just the
+    group name. E.g.:
+    1. ^(?P<a>\w+)/b/(\w+)$ ==> ^<a>/b/(\w+)$
+    2. ^(?P<a>\w+)/b/(?P<c>\w+)/$ ==> ^<a>/b/<c>/$
+    """
+    named_group_indices = [
+        (m.start(0), m.end(0), m.group(1))
+        for m in named_group_beg_matcher.finditer(pattern)
+    ]
+    group_pattern_and_name = []
+
+    # Loop over the found groups and their start, end indices
+    # Append named capture group pattern and its group name to a list.
+    for start, end, group_name in named_group_indices:
+        # Handle nested parentheses. E.g.: '^(?P<a>(x|y))/b'
+        unmatched_open_brackets, prev_char = 1, None
+        for idx, val in enumerate(list(pattern[end:])):
+            # If brackets are balanced, the end of the string for
+            # current named capture group pattern has been reached
+            if unmatched_open_brackets == 0:
+                group_pattern_and_name.append((pattern[start:end + idx], group_name))
+                break
+
+            # Check for unescaped `(` and `)`. They will mark start and end of a nested group
+            if val == '(' and prev_char != '\\':
+                unmatched_open_brackets += 1
+            elif val == ')' and prev_char != '\\':
+                unmatched_open_brackets -= 1
+            prev_char = val
+
+    # Replace string for named capture groups with their group names
+    for group_pattern, group_name in group_pattern_and_name:
+        pattern = pattern.replace(group_pattern, group_name)
+
+    return pattern
+
+
+def replace_non_named_groups(pattern):
+    """
+    Find unnamed groups in the pattern supplied, and replaces it '<var>'. E.g.:
+    1. ^(?P<a>\w+)/b/(\w+)$ ==> ^(?P<a>\w+)/b/<var>$
+    2. ^(?P<a>\w+)/b/((x|y)\w+)$ ==> ^(?P<a>\w+)/b/<var>$
+    """
+    group_name_default = "<var>"
+    unnamed_group_index = [m.start(0) for m in non_named_group_beg_matcher.finditer(pattern)]
```

Comment:
How about `unnamed_group_matcher = re.compile(r'\(')`? I can't make sense of the current complex regex. Test suite is passing with that. Can you edit this PR to allow me to push to your branch? If not I'll create a new PR.